community/kernel/src/main/java/org/neo4j/kernel/impl/transaction/log/EagerlyReversedTransactionCursor.java

```diff
@@ -0,0 +1,87 @@
+/*
+ * Copyright (c) 2002-2017 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+package org.neo4j.kernel.impl.transaction.log;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+
+import org.neo4j.kernel.impl.transaction.CommittedTransactionRepresentation;
+
+/**
+ * Eagerly exhausts a {@link TransactionCursor} and allows moving through it in reverse order.
+ * The idea is that this should only be done for a subset of a bigger transaction log stream, typically
+ * for one log file.
+ *
+ * For reversing a transaction log consisting of multiple log files {@link ReversedMultiFileTransactionCursor}
+ * should be used (it will use this class internally though).
+ *
+ * @see ReversedMultiFileTransactionCursor
+ */
+class EagerlyReversedTransactionCursor implements TransactionCursor
+{
+    private final List<CommittedTransactionRepresentation> txs = new ArrayList<>();
```

Comment:
Wouldn't a stack/linked list be more appropriate?

Reply:
More garbage tho, so I don't see the benefit really.

Reply:
Also, this class is a fallback for cases where the channel isn't a `ReadAheadLogChannel` so I'm not even sure this is ever needed. Could likely be removed at some point when/if we can verify that fact.

Reply:
`ArrayDeque` would have the same impact as an ArrayList, but you don't need the private int to keep track of index. But true, I could not find a case where this fallback would be used either.

Reply:
Right, `ArrayDeque` looks like a good candidate actually. But let's keep it as is since it will probably go away... I think it should.